DevTools/HackStudio/LanguageClient.h

```diff
@@ -0,0 +1,134 @@
+#include <DevTools/HackStudio/LanguageServers/LanguageServerEndpoint.h>
+#include <LibIPC/ServerConnection.h>
+
+namespace HackStudio {
+
+class LanguageClient;
+
+class ServerConnection
+    : public IPC::ServerConnection<LanguageClientEndpoint, LanguageServerEndpoint>
+    , public LanguageClientEndpoint {
+public:
+    ServerConnection(const StringView& socket, const StringView& project_path)
+        : IPC::ServerConnection<LanguageClientEndpoint, LanguageServerEndpoint>(*this, socket)
+        , m_project_path(project_path)
+    {
+    }
+
+    void attach(LanguageClient& client)
+    {
+        m_language_client = &client;
+    }
+
+    void detach()
+    {
+        m_language_client = nullptr;
+    }
+
+    virtual void handshake() override
+    {
+        auto response = send_sync<Messages::LanguageServer::Greet>(m_project_path.string());
+        set_my_client_id(response->client_id());
+    }
+
+    template<typename ConcreteType>
+    static NonnullRefPtr<ServerConnection> get_or_create(const String& project_path)
+    {
+        static HashMap<String, NonnullRefPtr<ConcreteType>> s_instances_for_projects;
+        auto key = LexicalPath { project_path }.string();
+        if (auto instance = s_instances_for_projects.get(key); instance.has_value())
+            return *instance.value();
+
+        auto connection = ConcreteType::construct(project_path);
+        connection->handshake();
+        s_instances_for_projects.set(key, *connection);
+        return *connection;
+    }
+
+protected:
+    virtual void handle(const Messages::LanguageClient::AutoCompleteSuggestions&) override;
+
+    LanguageClient* m_language_client { nullptr };
+    LexicalPath m_project_path;
+};
+
+class LanguageClient {
+public:
+    enum class AdoptTag {
+        Adopt,
+    };
+
+    explicit LanguageClient(ServerConnection& connection)
+        : m_connection(connection)
+    {
+        m_connection.attach(*this);
+    }
+
+    LanguageClient(NonnullRefPtr<ServerConnection>&& connection, AdoptTag)
+        : m_connection(*connection)
+        , m_server_connection(move(connection))
+    {
+        m_connection.attach(*this);
+    }
+
+    virtual ~LanguageClient()
+    {
+        m_connection.detach();
+    }
+
+    virtual void open_file(const String& path);
+    virtual void set_file_content(const String& path, const String& content);
+    virtual void insert_text(const String& path, const String& text, size_t line, size_t column);
+    virtual void remove_text(const String& path, size_t from_line, size_t from_column, size_t to_line, size_t to_column);
+    virtual void request_autocomplete(const String& path, size_t cursor_line, size_t cursor_column);
+
+    void provide_autocomplete_suggestions(const Vector<String>&);
+
+    Function<void(Vector<String>)> on_autocomplete_suggestions;
+
+private:
+    ServerConnection& m_connection;
+    // Populated only if this object owns the server connection.
+    RefPtr<ServerConnection> m_server_connection;
```

Comment:
Hmm this looks overly complicated :)
What if all `LanguageClient` objects simply have a `NonnullRefPtr<ServerConnection>` instead?

Reply:
🤔 I think the non-owning version isn't used anymore, so yeah, a NonnullRefPtr would work just fine.